File: solutions/c/pangram/1/pangram.c
#include "pangram.h"

#include <ctype.h>

bool is_pangram(const char* sentence) {
    if (!sentence) return false;

    bool letter_flags[26] = {0};
    for (const char* c = sentence; *c; ++c) {
        if (isalpha(*c)) letter_flags[tolower(*c) - 'a'] = true;
    }

    for (size_t i = 0; i < 26; ++i)
        if (!letter_flags[i]) return false;
    return true;
}

File: solutions/go/pangram/1/pangram.go
package pangram

import "unicode"

type unitType struct{}

var unit = unitType{}

type Set map[rune]unitType

func toSet(s string) Set {
	set := make(Set)
	for _, r := range s {
		set[unicode.ToLower(r)] = unit
	}
	return set
}

func isSubsetOf(seq string, set Set) bool {
	for _, r := range seq {
		if _, found := set[r]; !found {
			return false
		}
	}
	return true
}

var alphabet = "abcdefghijklmnopqrstuvwxyz"

func IsPangram(input string) bool {
	return isSubsetOf(alphabet, toSet(input))
}
